pkg/cmd/task/task_download.go

```diff
@@ -0,0 +1,174 @@
+package task
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/dcos/dcos-cli/api"
+	"github.com/dcos/dcos-core-cli/pkg/mesos"
+	"github.com/dcos/dcos-core-cli/pkg/pluginutil"
+	"github.com/spf13/cobra"
+)
+
+const description = `Downloads files from the sandbox of a given task.
+The <path> argument can specify a pattern for a single or multiple files.
+If no <path> is provided, the entire sandbox will be downloaded.`
+
+// newCmdMesosDownload downloads files from a tasks sandbox.
+func newCmdTaskDownload(ctx api.Context) *cobra.Command {
+	var targetDir string
+
+	cmd := &cobra.Command{
+		Use:   "download <task-id> [<path>]",
+		Short: description,
+		Args:  cobra.RangeArgs(1, 2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			task, err := findTask(ctx, args[0])
+			if err != nil {
+				return err
+			}
+
+			lastStatus := task.Statuses[0]
+			containerID := lastStatus.ContainerStatus.GetContainerID().Value
+
+			executorID := task.TaskID.Value
+			if task.ExecutorID != nil {
+				executorID = (*task.ExecutorID).Value
+			}
+
+			c := mesos.NewClient(pluginutil.HTTPClient(""))
+			paths, err := c.Debug(task.AgentID.Value)
+			if err != nil {
+				return err
+			}
+
+			path, err := getPath(paths, task.FrameworkID.Value, executorID, containerID)
+			if err != nil {
+				return err
+			}
+
+			pattern := "/"
+			if len(args) == 2 {
+				pattern = args[1]
+			}
+
+			return downloadDir(c, ctx, task.AgentID.Value, path, targetDir, pattern)
+		},
+	}
+
+	dir, err := os.Getwd()
+	if err == nil {
+		targetDir = dir
+	}
+	cmd.Flags().StringVar(&targetDir, "target-dir", targetDir, "Target directory of the download. Defaults to $PWD")
+	return cmd
+}
+
+func getPath(paths map[string]string, framework, executor, container string) (string, error) {
+	for k, v := range paths {
+		if strings.Contains(k, "/frameworks/"+framework+"/executors/"+executor+"/runs/"+container) {
+			return v, nil
+		}
+	}
+	return "", fmt.Errorf("unable to find task")
+}
+
+func downloadDir(c *mesos.Client, ctx api.Context, agentID, agentPath, targetDir, pattern string) error {
+	if err := ctx.Fs().MkdirAll(targetDir, 0755); err != nil {
+		return err
+	}
+
+	parent := parentPath(agentPath, pattern)
+	base := basePath(pattern)
+	files, err := c.Browse(agentID, parent)
+	if err != nil {
+		return err
+	}
+
+	errors := make(chan error, len(files))
+	for _, file := range files {
+		matched, err := matchFile(base, path.Base(file.Path))
+		if err != nil {
+			return fmt.Errorf("'%s' as pattern not supported: %v", pattern, err)
+		}
+
+		if !matched {
+			errors <- nil
+			continue
+		}
+
+		go func(f mesos.File) {
```

Comment:
maybe the content of the function could be extracted to a separate function for readability

Reply:
I first planned on doing that but the extraction created a function with a really long list or arguments to pass, so this looked cleaner to me.

Reply:
We could consider making a dedicated struct for this at some point.